bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java

```diff
@@ -0,0 +1,537 @@
+                    properties.put(PROPERTY_MODEL_ID, info.deviceModel);
+                    properties.put(PROPERTY_SERIAL_NUMBER, info.sn);
+                    properties.put(PROPERTY_FIRMWARE_VERSION, info.fwVersion);
+                    properties.put(PROPERTY_MAC_ADDRESS, info.wlanSTAMAC);
+                    updateStatus(ThingStatus.ONLINE);
+                }
+            } catch (JsonSyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            }
+        }
+    }
+
+    public void addChannel(String channelId, String itemType, @Nullable final Collection<String> options) {
+        if (thing.getChannel(channelId) == null) {
+            logger.trace("Channel '{}' for UID to be added", channelId);
+            ThingBuilder thingBuilder = editThing();
+            final ChannelTypeUID channelTypeUID = new ChannelTypeUID(BINDING_ID, channelId);
+            Channel channel = ChannelBuilder.create(new ChannelUID(getThing().getUID(), channelId), itemType)
+                    .withType(channelTypeUID).withKind(ChannelKind.STATE).build();
+            thingBuilder.withChannel(channel);
+            updateThing(thingBuilder.build());
+
+            if (options != null) {
+                final List<StateOption> stateOptions = options.stream()
+                        .map(e -> new StateOption(e, e.substring(0, 1) + e.toString().substring(1).toLowerCase()))
+                        .collect(Collectors.toList());
+                logger.trace("StateOptions : '{}'", stateOptions);
+                intesisStateDescriptionProvider.setStateOptions(channel.getUID(), stateOptions);
+            }
+        }
+    }
+
+    private void postRequest(UnaryOperator<String> requestFactory, Consumer<Response> handler) {
+        String sessionId = login();
+        if (sessionId != null) {
+            try {
+                String request = requestFactory.apply(sessionId);
+                logger.trace("request : '{}'", request);
+
+                String response = api.postRequest(ipAddress, request);
+                if (response != null) {
+
+                    Response resp = gson.fromJson(response, Response.class);
+                    boolean success = resp.success;
+                    if (success) {
+                        handler.accept(resp);
+                    } else {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                "Request unsuccessful");
+                    }
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No Response");
+                }
+            } catch (JsonSyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            } finally {
+                logout(sessionId);
+            }
+        }
+    }
```

Comment:
The previous suggestion requires a version of `postRequest` that doesn't handle session state. We can simply refactor some of the inner code from old `postRequest` method to serve this purpose.
But since we don't want to confuse our two `postRequest` methods, it would also be a good idea to
rename our old `postRequest` method into something more suitable. I suggest `postRequestInSession`.
```suggestion
    
    private void postRequest(String request, Consumer<Response> handler) {
        try {
                logger.trace("request : '{}'", request);

                String response = api.postRequest(ipAddress, request);
                if (response != null) {

                    Response resp = gson.fromJson(response, Response.class);
                    boolean success = resp.success;
                    if (success) {
                        handler.accept(resp);
                    } else {
                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                                "Request unsuccessful");
                    }
                } else {
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No Response");
                }
         } catch (JsonSyntaxException e) {
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
         }
    }

    private void postRequestInSession(UnaryOperator<String> requestFactory, Consumer<Response> handler) {
        String sessionId = login();
        if (sessionId != null) {
            try {
                String request = requestFactory.apply(sessionId);
                postRequest(request, handler);
            } finally {
                logout(sessionId);
            }
        }
    }
```

Reply:
Done